File: samples/snippets/fsharp/VS_Snippets_CLR_System/system.Action~2/fs/Delegate.fs
module TestDelegate

// <Snippet1>
open System
open System.IO

type ConcatStrings = delegate of string1: string * string1: string -> unit

let message1 = "The first line of a message"
let message2 = "The second line of a message"

let writeToConsole string1 string2 = 
    printfn $"{string1}\n{string2}"

let writeToFile string1 string2 =
    use writer = new StreamWriter(Environment.GetCommandLineArgs().[1], false)
    try
        writer.WriteLine $"{string1}\n{string2}"
    with _ -> printfn "File write operation failed..."

let concat =
    ConcatStrings(
        if Environment.GetCommandLineArgs().Length > 1 then
            writeToFile
        else
            writeToConsole
    )

concat.Invoke(message1, message2)

// </Snippet1>

File: snippets/csharp/System/ActionT1,T2/Overview/Delegate.cs
// <Snippet1>
using System;
using System.IO;

delegate void ConcatStrings(string string1, string string2);

public class TestDelegate
{
    public static void Main()
    {
        string message1 = "The first line of a message.";
        string message2 = "The second line of a message.";
        ConcatStrings concat;

        if (Environment.GetCommandLineArgs().Length > 1)
            concat = WriteToFile;
        else
            concat = WriteToConsole;

        concat(message1, message2);
    }

    private static void WriteToConsole(string string1, string string2) => Console.WriteLine($"{string1}\n{string2}");

    private static void WriteToFile(string string1, string string2)
    {
        StreamWriter writer = null;
        try
        {
            writer = new(Environment.GetCommandLineArgs()[1], false);
            writer.WriteLine($"{string1}\n{string2}");
        }
        catch
        {
            Console.WriteLine("File write operation failed...");
        }
        finally
        {
            if (writer != null) writer.Close();
        }
    }
}
// </Snippet1>
